### DISCUSSION TITLE
wifi: support 802.11k and 802.11v

### DISCUSSION THREAD

**Discussion by nagyrobi** [HASwitchPlate/openHASP#347] [Ideas]
Do we support these?
Idea from: https://github.com/esphome/esphome/pull/3600

**Comment by fvanroie** [comment]
Support for these Wifi features were added in ESP-IDF 4.4. I see that the flags are present in Arduino 2.0 but are not exposed.
The `Wifi.connect()` method does not allow these parameters to be set yet from Arduino.

Either the flags need to be added to the Arduino Wifi library, or a custom connect method needs to be implemented like ESPhome...

**Reply by fvanroie** [reply]
The support for these are very new. I found a roaming example for ESP-IDF: [roaming_example.c](https://github.com/espressif/esp-idf/blob/ee33c6ae05d55acbc462afd8595b432d9002e62d/examples/wifi/roaming/main/roaming_example.c)
However, some of the functions used are not yet in the latest [Tasmota Core 2.0.4pre](https://github.com/tasmota/platform-espressif32/releases/tag/v2.0.4pre) with ESP-IDF v4.4.1.1.

From what I gather RRM (802.11k) might be possible without custom code, but BTM (802.11v) requires application code to activate re-scanning and select an alternate AP when the RRSI drops below a certain threshold. I would expect this could/should be handled best by the Arduino Core, so all sketches can benefit from this feature... But Arduino Core is lacking this atm.

**Reply by nagyrobi** [reply]
Understood, thanks for looking after.

**Comment by fvanroie** [comment]
I added a `feature/wifi` branch that enables both flags via the ESP-IDF commands. Not sure if this is enough, but the WiFi stack reports them as enabled now (disabled by default):
```
[       4.107][86004/100272 14][53128/53132  1] WIFI: Connected to test, requesting IP...
[       4.129][86004/99692 13][53128/53132  1] WIFI: BTM: Enabled
[       4.141][86004/100032 14][53128/53132  1] WIFI: RM : Enabled
[       4.153][86004/100180 14][53128/53132  1] WIFI: Received IP address 192.168.1.123
[       4.167][86004/100180 14][53128/53132  1] ----: Connected = online
```
I don't know how to test/verify the features properly, so you might want to do some testing in your network first.
